### PR TITLE
hubble: add separate API to get agent and debug events

### LINKED ISSUE #15111: hubble: Clean up API for debug and agents events
Cilium 1.10 adds support for debug and agent events in the Hubble API. They are currently exposed as part of the `GetFlows` API:

https://github.com/cilium/cilium/blob/676c2add9849bd9ec00448c7d0b13c2ed6d73736/api/v1/observer/observer.proto#L119-L123

Before we ship this API in Cilium 1.10, we want to split these out into separate APIs, e.g. `GetAgentEvents`, and `GetDebugEvents`, as the current mixing of types in `GetFlows` events poses issues with filtering.

This is a release blocker, because we do not want to commit to the current API present in `master`.

### PR DESCRIPTION
**Pull request by tklauser**
Review by commit:

- Commits 1-3 are preparatory refactoring to simplify/re-use code in the successive commits.
- The 4th commit adds the new API but does not implement it yet. I split this from the following commit to separate the big chunk of regenerated protobuf API.
- The 5th commit implements the new `GetAgentEvents` and `GetDebugEvents` API.
- The 6th commit removes the agent and debug events from the existing `GetFlows` API.

See commit messages for more details.

Notes:
- This does not yet implement filtering or block-/allowlisting of events as in the `GetFlows` API.
- The API is currently stubbed out in Hubble Relay. This will be implemented in a follow-up PR.

Fixes: #15111

### CODE REVIEW

**Comment by tklauser** [comment]
test-me-please

File: pkg/hubble/observer/local_observer.go [hunk 1]
```diff
@@ -328,43 +492,44 @@ type eventsReader struct {
...
-		until := req.Until.AsTime()
-		reader.until = &until
+		untilTime := until.AsTime()
+		reader.until = &untilTime
 	}
 
 	return reader, nil
 }
 
 // Next returns the next flow that matches the request criteria.
```
Comment (kaworu) [inline pkg/hubble/observer/local_observer.go]:
```suggestion
// Next returns the next event matching the request criteria.
```

File: pkg/hubble/testutils/fake.go [hunk 2]
```diff
@@ -52,12 +52,30 @@ func (s *FakeGetFlowsServer) Send(response *observerpb.GetFlowsResponse) error {
...
+// observerpb.Observer_GetAgentEventsServer interface.
+type FakeGetAgentEventsServer struct {
+	OnSend func(response *observerpb.GetAgentEventsResponse) error
+	*FakeGRPCServerStream
+}
+
+// Send implements observerpb.Observer_GetAgentEventsServer.Send.
+func (s *FakeGetAgentEventsServer) Send(response *observerpb.GetAgentEventsResponse) error {
+	if s.OnSend != nil {
+		// TODO: completely convert this into using flowpb.Flow
```
Comment (kaworu) [inline pkg/hubble/testutils/fake.go]: I don't think this comment apply here

**Comment by tklauser** [comment]
test-gke

File: pkg/hubble/observer/local_observer.go [hunk 3]
```diff
@@ -306,6 +333,128 @@ nextEvent:
...
+		log.WithFields(logrus.Fields{
+			"number_of_agent_events": i,
+			"buffer_size":            ring.Cap(),
+			"took":                   time.Since(start),
+		}).Debug("GetAgentEvents finished")
+	}()
+
+	ringReader, err := newRingReader(ring, req, whitelist, blacklist)
+	if err != nil {
+		if err == io.EOF {
```
Comment (kaworu) [inline pkg/hubble/observer/local_observer.go]: nitpicking:
```suggestion
		if errors.Is(err, io.EOF) {
```
Could do the same below and in `GetFlows`.

**Comment by tklauser** [comment]
test-1.16-netnext

File: api/v1/observer/observer.proto [hunk 4]
```diff
@@ -174,3 +230,25 @@ message TLS {
     // cert validation process.
     string server_name = 2;
 }
+
+// ExportEvent contains an event to be exported.
```
Comment (kaworu) [inline api/v1/observer/observer.proto]: Can we add a warning comment that this protobuf message is not intended to be used outside of the export feature please?

**Comment by tklauser** [comment]
test-1.20-4.19

**Review by kaworu — APPROVED** [review]
awesome work @tklauser thanks a lot for breaking this PR into commits that were easy to review! I have some comments and nits but the patch LGTM.

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-1.16-netnext

**Review by gandro — APPROVED** [review]
Looks good! Thanks a ton for taking care of this!

**Review by nathanjsweet — APPROVED** [review]

**Comment by tklauser** [comment]
test-1.16-netnext

**Review by rolinh — APPROVED** [review]
LGTM, thanks!

**Pull request merged**